pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+        # initial the identity is used
+        self._B = np.identity(self._n_parameters)
+
+        # newton direction
+        self.__px = None
+
+        # number of accepted steps/ newton direction updates
+        self._k = 0
+
+        # number of steps in the current line search iteration
+        self.__j = 0
+        self.__logic_steps_left = 0
+
+        # maximum number of line search steps before a successful point
+
+        # packed reply of objective function and gradient evaluations
+        # needed by generator implementation
+        # this approach is taken rather than the send() keyword to
+        # back date past Python 2.5
+        self.__reply_f_and_dfdx = None
+
+        # Current point, score, and gradient
+        self._current = self._x0
+        self._current_f = None
+        self._current_dfdx = None
+
+        # Proposed next point (read-only, so can be passed to user)
+        self._proposed = self._x0
+        self._proposed.setflags(write=False)
+
+        self.__convergence = False
+
+        # ask generator
+        self.__ask_generator = self.__ask_generator()
+
+        # need to check wolfe conditions
+        self.wolfe_check = False
+        # continue main body of line search loop
+        self.__line_search_looping = None
+        # wheather the wolfe conditions have been meet
+        self.__wolfe_cond_meet = False
+
+    def name(self):
+        """ See :meth:`Optimiser.name()`. """
+        return 'Hager-Zhang Line Search'
+
+    def fbest(self):
+        """ See :meth:`Optimiser.fbest()`. """
+        return self._fbest
+
+    def wolfe_line_search_parameters(self):
+        """
+        Returns the  wolfe line search parameters this optimiser is using
+        as a vector ``[c1, c2]``.
+        As c1 approaches 0 and c2 approaches 1, the line search terminates
+        more quickly. ``c1`` is the parameter for the Armijo condition
+        rule, ``0 < c1 < 0.5``. ``c2 `` is the parameter for the
+        curvature condition rule, ``c1 < c2 < 1.0``.
+        """
+        return (self._c1, self._c2)
+
+    def needs_sensitivities(self):
+        """ See :meth:`Optimiser.needs_sensitivities()`. """
+        return True
+
+    def n_hyper_parameters(self):
+        """ See :meth:`pints.TunableMethod.n_hyper_parameters()`. """
+        return 2
+
+    def running(self):
+        """ See :meth:`Optimiser.running()`. """
+        return self._running
+
+    def set_hyper_parameters(self, x):
+        """
+        See :meth:`pints.TunableMethod.set_hyper_parameters()`.
+
+        The hyper-parameter vector is ``[c1, c2, m]``.
+        ``c1`` is the parameter for the Armijo condition rule,
+        ``0 < c1 < 0.5``.
+        ``c2`` is the parameter for the curvature condition rule,
+        ``c1 < c2 < 1.0``.
+        ``m`` is the number of maximum number of correction matrices
+        that can be stored for the LM-BFGS update.
+        """
+
+        self.__set_wolfe_line_search_parameters(x[0], x[1])
```

Comment:
There seem to be three parameters mentioned in the docstrings but only two set here? Or is one of them a vector?